Update ENTRYPOINT/CMD table to agree with docs

ddgenome opened the pull request:
**\- What I did**

In the Dockerfile reference documentation, the table describing the interaction of ENTRYPOINT and CMD was not in agreement with what other parts of the document said. For example, this sentence https://github.com/docker/docker/blame/master/docs/reference/builder.md#L1084-L1085 states that CMD is ignored when the shell form of the ENTRYPOINT command is used but the table indicated otherwise. I tested myself which was correct, finding the table was incorrect, so I updated the table.

**\- How I did it**

I edited the markdown in a text editor.

**\- How to verify it**

You can create a Dockerfile with the following contents:

```
FROM busybox:latest
ENTRYPOINT /bin/echo p1_entry
CMD shell_cmd p1_cmd
```

build a container, run it, and see that `p1_entry` is output but `shell_cmd p1_cmd` is not.  Similar for an exec form of CMD.

**\- Description for the changelog**

Dockerfile documentation more consistent and accurate

**\- A picture of a cute animal (not mandatory but encouraged)**

![Thumbs up](http://justcuteanimals.com/wp-content/uploads/2014/07/hamster-thumbs-up-cute-hamters-animal-picture-pics.jpg)

Several other places in the document it states that when using the shell
form of ENTRYPOINT, CMD and command line arguments are ignored.  That is
accurate, this table was not.  It is now.

Signed-off-by: David Dooling dooling@gmail.com

ddgenome committed:
Update ENTRYPOINT/CMD table to agree with docs

Several other places in the document it states that when using the shell
form of ENTRYPOINT, CMD and command line arguments are ignored.  That is
accurate, this table was not.  It is now.

Signed-off-by: David Dooling <dooling@gmail.com>

cpuguy83 commented:
ping @sfsmithcha

sfsmithcha commented:
I think you are right that this table is a little confusing. However I don't think that your update is correct.

The way I understand it, the column **ENTRYPOINT exec_entry p1_entry** is meant to show how the container executes a combination of that `ENTRYPOINT` with the `CMD` from the first column.

I could be wrong, but I'm not sure your example test accurate. I think `shell_cmd`  would need to be a valid command in `busybox`:

```
$ docker run -it --rm busybox
/ # shell_cmd
sh: shell_cmd: not found
```

cc: @cpuguy83 , @vdemeester , @mstanleyjones

ddgenome commented:
@sfsmithcha you are correct that the table is meant to show how the container executes a combination of an ENTRYPOINT with a CMD. Unfortunately, the table is wrong. When a shell form of ENTRYPOINT is used, CMD is ignored, as stated several other places in the document. That is why my patch removes the specific text it does, as it contradicts the other places in the document and the actual behavior.

`shell_cmd` is not meant to be executed alone, it is meant as an argument provided to the ENTRYPOINT, which in my example is `echo`. It is meant to be a place holder for a command or text that someone would provide either in a CMD Dockerfile command or on the command-line. In either case, it is immaterial since it is ignored when the shell form of ENTRYPOINT is used.

If you create the Dockerfile I describe above and run it you will see that `p1_entry` is output but `shell_cmd p1_cmd` is not. If you prefer, replace `shell_cmd` with and actual command like `ls` and you'll see the same response.

```sh
$ docker version
Client:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Fri Feb 12 22:11:40 UTC 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   f1e1b83
 Built:
 OS/Arch:      linux/amd64
$ cat Dockerfile
FROM busybox:latest
ENTRYPOINT /bin/echo p1_entry
CMD shell_cmd p1_cmd
$ docker build .
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM busybox:latest
 ---> 2b8fd9751c4c
Step 2 : ENTRYPOINT /bin/echo p1_entry
 ---> Running in ecb0f279bc43
 ---> 5c380b4d10b6
Removing intermediate container ecb0f279bc43
Step 3 : CMD exec_cmd p1_cmd
 ---> Running in 46c4d9e020e9
 ---> 89d6415ec541
Removing intermediate container 46c4d9e020e9
Successfully built 89d6415ec541
$ docker run --rm 89d6415ec541
p1_entry
$ docker run --rm 89d6415ec541 foo bar
p1_entry
$ cat DockerfileExec
FROM busybox:latest
ENTRYPOINT /bin/echo p1_entry
CMD ["exec_cmd", "p1_cmd"]
$ docker build -f DockerfileExec  .
Sending build context to Docker daemon 3.072 kB
Step 1 : FROM busybox:latest
 ---> 2b8fd9751c4c
Step 2 : ENTRYPOINT /bin/echo p1_entry
 ---> Running in 701e0215e830
 ---> 102919a899bd
Removing intermediate container 701e0215e830
Step 3 : CMD exec_cmd p1_cmd
 ---> Running in 091e99cab0b6
 ---> 9bc311a02243
Removing intermediate container 091e99cab0b6
Successfully built 9bc311a02243
$ docker run --rm 9bc311a02243
p1_entry
$ docker run --rm 9bc311a02243 foo bar
p1_entry
```

sfsmithcha commented:
@mstanleyjones PTAL

SvenDowideit commented:
LGTM

mstanleyjones commented:
Actually, italic is usually used for replaceable text. You'd probably want italic monotext. It's a little tricky in markdown, but use an underscore followed by the backtick followed by the phrase. Like _`this`_.

ddgenome commented:
@SvenDowideit and @mstanleyjones you seem to be commenting on the format of the comments in the PR and not the actual PR, in which I changed no formatting and added no text. I only removed some erroneous text from the table.

SvenDowideit commented:
@ddgenome nope - we're suggesting improving how the existing text could be improved.

sadly, I had an escaping fail - I meant to say put `<` and `>` around the non-functional examples

ddgenome commented:
@SvenDowideit The desire to change how the table data is formatted is orthogonal to fixing the erroneous data in the table. Can this be merged in and an issue opened for your desired changes around formatting?
